### PR TITLE
Columns Block: Don't show the column count change UI when `templateLock` is `all`

### PR DESCRIPTION
**Pull request by t-hamano**
See: https://github.com/WordPress/gutenberg/issues/48620#issuecomment-1451257024

## What?
This PR hides the column change UI in the sidebar if the `core/columns` block has `templateLock=all`, i.e., the `core/column` block is not allowed to be inserted.

## Why?
If `templateLock` is `all`, no inner blocks are allowed to be inserted. However, by using the slider to change the number of columns, it is possible to insert columns that are in fact inner blocks. Given the role of templateLock, I think it should not be the intent that the UI can be manipulated.

## How?
I used `canInsertBlockType()` to not show the column change UI if the `core/columns` block is in a context where it cannot be inserted.

## Testing Instructions

Use the following HTML to insert a group block with `templateLock=all` set.

```html
<!-- wp:group {"templateLock":"all","layout":{"type":"constrained"}} -->
<div class="wp-block-group"><!-- wp:columns -->
<div class="wp-block-columns"><!-- wp:column -->
<div class="wp-block-column"><!-- wp:paragraph -->
<p>Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.</p>
<!-- /wp:paragraph --></div>
<!-- /wp:column -->

<!-- wp:column -->
<div class="wp-block-column"><!-- wp:paragraph -->
<p>Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.</p>
<!-- /wp:paragraph --></div>
<!-- /wp:column --></div>
<!-- /wp:columns --></div>
<!-- /wp:group -->
```

Confirm that the column count change UI is not displayed when selecting a `core/columns` block in a group block.

### Before

before

### After

after

### CODE REVIEW

**Review by andrewserong — APPROVED** [review]
Thanks for the ping @t-hamano, this PR's testing nicely for me for the templateLock `all` case, while not causing any unrelated issues for the Column(s) blocks 👍

One thing I noticed in testing is that there's another case that might be good to try to handle, which is when blocks are locked via the `lock` attribute or via the UI for locking. For example, if I go to lock individual column blocks against removal, but then move up to the parent Columns block, then I can still adjust the slider to remove the column blocks:

https://user-images.githubusercontent.com/14988353/222594808-41e9b1f3-c970-4278-99f5-cbb5d55a3023.mp4

This isn't strictly speaking related to `templateLock`, though, and the fix in this PR still looks like a good approach to me, so I don't think this is a blocker.

This LGTM! ✨

**Comment by t-hamano** [comment]
Thank you for the review, @andrewserong!

> One thing I noticed in testing is that there's another case that might be good to try to handle, which is when blocks are locked via the lock attribute or via the UI for locking. For example, if I go to lock individual column blocks against removal, but then move up to the parent Columns block, then I can still adjust the slider to remove the column blocks:

This is a bit of a complicated issue. Because we also need to consider the case where only some columns are locked.

For example, only one of the four columns is locked, as shown below:

- Column 1
- Column 2
- Column 3 (Locked)
- Column 4

If the number of columns is reduced, the columns are removed from the back of the list. If I change the number of columns from **4 to 2** in the text control, I would expect the following layout:

- Column 1
- Column 2
- Column 3 (Locked)

In other words, the logic is to delete columns in order from the back, and if a locked column is hit, no more columns are deleted.

I will try to resolve this issue in another PR 👍
